// hypixel-api-reborn/dist/Structures/Boosters/Booster.js
import Game from '../Game.js';
export function parseType(data) {
    if (true === data.stacked)
        return 'STACKED';
    if (!data.stacked)
        return 'QUEUED';
    return 'ACTIVE';
}
class Booster {
    purchaser;
    amount;
    originalLength;
    remaining;
    activatedTimestamp;
    activated;
    game;
    isActive;
    type;
    stackers;
    expired;
    constructor(data) {
        this.purchaser = data.purchaserUuid;
        this.amount = data.amount;
        this.originalLength = data.originalLength;
        this.remaining = data.length;
        this.activatedTimestamp = data.dateActivated;
        this.activated = new Date(data.dateActivated);
        this.game = new Game(data.gameType);
        this.isActive = Array.isArray(data.stacked);
        this.type = parseType(data);
        this.stackers = Array.isArray(data.stacked) ? Array.from(data.stacked) : [];
        this.expired = 0 > data.length;
    }
    toString() {
        return `${this.purchaser}'s booster in ${this.game}`;
    }
}
export default Booster;
