// farming-weight/dist/constants/lotus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GREEN_THUMB_MULTIPLIER = exports.VISITORS_SERVED_BONUS = exports.EQUIPMENT_INFO = void 0;
const armor_1 = require("./armor");
const reforges_1 = require("./reforges");
const skills_1 = require("./skills");
const upgrades_1 = require("./upgrades");
exports.EQUIPMENT_INFO = {
    LOTUS_BRACELET: {
        name: 'Lotus Bracelet',
        wiki: 'https://wiki.hypixel.net/Lotus_Bracelet',
        family: 'LOTUS',
        slot: armor_1.GearSlot.Gloves,
        maxRarity: reforges_1.Rarity.Epic,
        stats: {
            [reforges_1.Stat.FarmingFortune]: 5,
        },
        skillReq: {
            [skills_1.Skill.Farming]: 10,
            [skills_1.Skill.Garden]: 2,
        },
    },
    LOTUS_BELT: {
        name: 'Lotus Belt',
        wiki: 'https://wiki.hypixel.net/Lotus_Belt',
        family: 'LOTUS',
        slot: armor_1.GearSlot.Belt,
        maxRarity: reforges_1.Rarity.Epic,
        stats: {
            [reforges_1.Stat.FarmingFortune]: 5,
        },
        skillReq: {
            [skills_1.Skill.Farming]: 10,
            [skills_1.Skill.Garden]: 3,
        },
    },
    LOTUS_NECKLACE: {
        name: 'Lotus Necklace',
        wiki: 'https://wiki.hypixel.net/Lotus_Necklace',
        family: 'LOTUS',
        slot: armor_1.GearSlot.Necklace,
        maxRarity: reforges_1.Rarity.Epic,
        stats: {
            [reforges_1.Stat.FarmingFortune]: 5,
        },
        skillReq: {
            [skills_1.Skill.Farming]: 10,
            [skills_1.Skill.Garden]: 4,
        },
    },
    LOTUS_CLOAK: {
        name: 'Lotus Cloak',
        wiki: 'https://wiki.hypixel.net/Lotus_Cloak',
        upgrade: {
            id: 'ZORROS_CAPE',
            reason: upgrades_1.UpgradeReason.Situational,
            why: 'A maxed Zorro\'s Cloak provides significantly more fortune during a Jacob\'s contest, but slightly less outside of one.',
        },
        family: 'LOTUS',
        slot: armor_1.GearSlot.Cloak,
        maxRarity: reforges_1.Rarity.Epic,
        stats: {
            [reforges_1.Stat.FarmingFortune]: 5,
        },
        skillReq: {
            [skills_1.Skill.Farming]: 10,
            [skills_1.Skill.Garden]: 5,
        },
    },
    ZORROS_CAPE: {
        name: "Zorro's Cloak",
        wiki: 'https://wiki.hypixel.net/Zorro%27s_Cape',
        upgrade: {
            id: 'LOTUS_CLOAK',
            reason: upgrades_1.UpgradeReason.Situational,
            why: 'A maxed Zorro\'s Cape provides significantly more fortune during a Jacob\'s contest, but slightly less outside of one compared to a maxed Lotus Cloak.',
        },
        slot: armor_1.GearSlot.Cloak,
        maxRarity: reforges_1.Rarity.Mythic,
        stats: {
            [reforges_1.Stat.FarmingFortune]: 10,
            [reforges_1.Stat.FarmingWisdom]: 1,
            [reforges_1.Stat.Strength]: 10,
            [reforges_1.Stat.Ferocity]: 2,
        },
        contestStatsMultiplier: 2
    },
};
exports.VISITORS_SERVED_BONUS = {
    1: 1,
    5: 2,
    10: 3,
    20: 4,
    50: 5,
    75: 6,
    100: 7,
    150: 8,
    250: 9,
    500: 10,
    1000: 11,
    2000: 12,
    3000: 13,
    4000: 14,
    5000: 15,
};
exports.GREEN_THUMB_MULTIPLIER = {
    1: 0.05,
    2: 0.1,
    3: 0.15,
    4: 0.2,
    5: 0.25,
};
