// farming-weight/dist/constants/specialcrops.js
import { Crop } from './crops.js';
export var SpecialCrop;
(function (SpecialCrop) {
    SpecialCrop["Cropie"] = "Cropie";
    SpecialCrop["Squash"] = "Squash";
    SpecialCrop["Fermento"] = "Fermento";
})(SpecialCrop || (SpecialCrop = {}));
export const SPECIAL_CROP_INFO = {
    [SpecialCrop.Cropie]: {
        name: 'Cropie',
        npc: 25000,
        rates: [0.0003, 0.0003, 0.0004, 0.0005],
    },
    [SpecialCrop.Squash]: {
        name: 'Squash',
        npc: 75000,
        rates: [0.0001, 0.0001, 0.0002, 0.0003],
    },
    [SpecialCrop.Fermento]: {
        name: 'Fermento',
        npc: 250000,
        rates: [0.00005, 0.00005, 0.00006, 0.00007],
    },
};
export const MATCHING_SPECIAL_CROP = {
    [Crop.Wheat]: SpecialCrop.Cropie,
    [Crop.Carrot]: SpecialCrop.Cropie,
    [Crop.Potato]: SpecialCrop.Cropie,
    [Crop.CocoaBeans]: SpecialCrop.Squash,
    [Crop.Melon]: SpecialCrop.Squash,
    [Crop.Pumpkin]: SpecialCrop.Squash,
    [Crop.Cactus]: SpecialCrop.Fermento,
    [Crop.Mushroom]: SpecialCrop.Fermento,
    [Crop.NetherWart]: SpecialCrop.Fermento,
    [Crop.SugarCane]: SpecialCrop.Fermento,
    [Crop.Seeds]: SpecialCrop.Fermento,
};
