// farming-weight/dist/constants/stats.js
export var Stat;
(function (Stat) {
    Stat["Strength"] = "Strength";
    Stat["Health"] = "Health";
    Stat["Defense"] = "Defense";
    Stat["Speed"] = "Speed";
    Stat["Intelligence"] = "Intelligence";
    Stat["CritChance"] = "Crit Chance";
    Stat["CritDamage"] = "Crit Damage";
    Stat["AttackSpeed"] = "Attack Speed";
    Stat["AbilityDamage"] = "Ability Damage";
    Stat["MagicFind"] = "Magic Find";
    Stat["PetLuck"] = "Pet Luck";
    Stat["TrueDefense"] = "True Defense";
    Stat["SeaCreatureChance"] = "Sea Creature Chance";
    Stat["Ferocity"] = "Ferocity";
    Stat["MiningSpeed"] = "Mining Speed";
    Stat["MiningFortune"] = "Mining Fortune";
    Stat["FarmingFortune"] = "Farming Fortune";
    Stat["ForagingFortune"] = "Foraging Fortune";
    Stat["MiningWisdom"] = "Mining Wisdom";
    Stat["FarmingWisdom"] = "Farming Wisdom";
    Stat["ForagingWisdom"] = "Foraging Wisdom";
    Stat["Pristine"] = "Pristine";
    Stat["BonusPestChance"] = "Bonus Pest Chance";
})(Stat || (Stat = {}));
export function getStatValue(stat, option) {
    if (!stat || (stat.exists && option && !stat.exists(option)))
        return 0;
    let value = 0;
    if ('value' in stat) {
        value += stat.value;
    }
    if ('calculated' in stat && option) {
        value += stat.calculated(option);
    }
    return value;
}
