// farming-weight/dist/constants/upgrades.js
export var UpgradeReason;
(function (UpgradeReason) {
    UpgradeReason["NextTier"] = "next";
    UpgradeReason["DeadEnd"] = "dead";
    UpgradeReason["Situational"] = "situational";
})(UpgradeReason || (UpgradeReason = {}));
export var UpgradeCategory;
(function (UpgradeCategory) {
    UpgradeCategory["Enchant"] = "enchant";
    UpgradeCategory["Rarity"] = "rarity";
    UpgradeCategory["Item"] = "item";
    UpgradeCategory["Gem"] = "gem";
})(UpgradeCategory || (UpgradeCategory = {}));
export var UpgradeAction;
(function (UpgradeAction) {
    UpgradeAction["Apply"] = "apply";
    UpgradeAction["Recombobulate"] = "recombobulate";
    UpgradeAction["LevelUp"] = "levelup";
})(UpgradeAction || (UpgradeAction = {}));
// export const INITIAL_ITEMS: InitialItems = {
// 	tools: {
// 		[Crop.Cactus]: 'CACTUS_KNIFE',
// 		[Crop.CocoaBeans]: 'COCO_CHOPPER',
// 		[Crop.Carrot]: 'THEORETICAL_HOE_CARROT_1',
// 		[Crop.Melon]: 'MELON_DICER',
// 		[Crop.Mushroom]: 'FUNGI_CUTTER',
// 		[Crop.NetherWart]: 'THEORETICAL_HOE_WARTS_1',
// 		[Crop.Potato]: 'THEORETICAL_HOE_POTATO_1',
// 		[Crop.Pumpkin]: 'PUMPKIN_DICER',
// 		[Crop.SugarCane]: 'THEORETICAL_HOE_CANE_1',
// 		[Crop.Wheat]: 'THEORETICAL_HOE_WHEAT_1'
// 	},
// 	armor: {
// 		[GearSlot.Boots]: [ 'MELON_BOOTS', 'FARMER_BOOTS', 'RANCHERS_BOOTS' ],
// 		[GearSlot.Leggings]: 'MELON_LEGGINGS',
// 		[GearSlot.Chestplate]: 'MELON_CHESTPLATE',
// 		[GearSlot.Helmet]: 'MELON_HELMET',
// 		[GearSlot.Necklace]: 'LOTUS_NECKLACE',
// 		[GearSlot.Cloak]: 'LOTUS_CLOAK',
// 		[GearSlot.Belt]: 'LOTUS_BELT',
// 		[GearSlot.Gloves]: 'LOTUS_BRACELET',
// 	},
// 	pets: [ 'ELEPHANT', 'MOOSHROOM_COW' ]
// };
