// hypixel-api-reborn/dist/Structures/Housing.js
class HousingGivenCookies {
    date;
    houses;
    constructor(date, houses) {
        this.date = date;
        this.houses = houses;
    }
}
class Housing {
    allowedBlocks;
    playerVisibility;
    showBorder;
    showTips;
    tutorialStage;
    packages;
    firstHouseJoinTimestamp;
    firstHouseJoinAt;
    plotSize;
    givenCookies;
    constructor(data) {
        this.allowedBlocks = data.allowedBlocks;
        this.playerVisibility = data?.playerSettings?.VISIBILITY
            ? Number(data?.playerSettings?.VISIBILITY.split('-')[1])
            : 4;
        this.showBorder = data?.playerSettings?.BORDER ? Boolean(data?.playerSettings?.BORDER.split('-')[1]) : true;
        this.showTips = data?.playerSettings?.TIPS ? Boolean(data?.playerSettings?.TIPS.split('-')[1]) : false;
        this.tutorialStage = data.tutorialStep;
        this.packages = data.packages;
        this.firstHouseJoinTimestamp = data.firstHouseJoinMs;
        this.firstHouseJoinAt = new Date(this.firstHouseJoinTimestamp);
        this.plotSize = data.plotSize;
        this.givenCookies = [];
        Object.keys(data)
            .filter((key) => key.startsWith('given_cookies_'))
            .map((key) => {
            this.givenCookies.push(new HousingGivenCookies(key, data[key]));
        });
    }
}
export default Housing;
