// hypixel-api-reborn/dist/Structures/MiniGames/BlitzSurvivalGames/BlitzSurvivalGames.js
import BlitzSurvivalGamesKit from './BlitzSurvivalGamesKit.js';
import Divide from '../../../Utils/Divide.js';
class BlitzSurvivalGames {
    coins;
    kills;
    kit;
    killsSolo;
    killsTeams;
    deaths;
    KDR;
    wins;
    winsSolo;
    winsTeam;
    gamesPlayed;
    losses;
    WLR;
    arrowsShot;
    arrowsHit;
    bowAccuracy;
    damage;
    damageTaken;
    potionsDrunk;
    potionsThrown;
    mobsSpawned;
    playTime;
    blitzUses;
    chestsOpened;
    archer;
    meatmaster;
    speleologist;
    baker;
    knight;
    guardian;
    scout;
    hunter;
    hypeTrain;
    fisherman;
    armorer;
    horsetamer;
    astronaut;
    troll;
    reaper;
    shark;
    reddragon;
    toxicologist;
    rogue;
    warlock;
    slimeyslime;
    jockey;
    golem;
    viking;
    shadowKnight;
    pigman;
    paladin;
    necromancer;
    florist;
    diver;
    arachnologist;
    blaze;
    wolftamer;
    tim;
    farmer;
    creepertamer;
    snowman;
    constructor(data) {
        this.coins = data?.coins || data?.tokens || 0;
        this.kills = data?.kills || 0;
        this.kit = data?.defaultkit || 'None';
        this.killsSolo = data?.kills_solo_normal || 0;
        this.killsTeams = data?.kills_teams_normal || 0;
        this.deaths = data?.deaths || 0;
        this.KDR = Divide(this.kills, this.deaths);
        this.wins = data?.wins || 0;
        this.winsSolo = data?.wins_solo_normal || 0;
        this.winsTeam = data?.wins_teams || 0;
        this.gamesPlayed = data?.games_played || 0;
        this.losses = this.gamesPlayed - this.wins;
        this.WLR = Divide(this.wins, this.losses);
        this.arrowsShot = data?.arrows_fired || 0;
        this.arrowsHit = data?.arrows_hit || 0;
        this.bowAccuracy = Divide(this.arrowsHit, this.arrowsShot);
        this.damage = data?.damage || 0;
        this.damageTaken = data?.damage_taken || 0;
        this.potionsDrunk = data?.potions_drunk || 0;
        this.potionsThrown = data?.potions_thrown || 0;
        this.mobsSpawned = data?.mobs_spawned || 0;
        this.playTime = data?.time_played || 0;
        this.blitzUses = data?.blitz_uses || 0;
        this.chestsOpened = data?.chests_opened || 0;
        this.archer = new BlitzSurvivalGamesKit(data, 'archer');
        this.meatmaster = new BlitzSurvivalGamesKit(data, 'meatmaster');
        this.speleologist = new BlitzSurvivalGamesKit(data, 'speleologist');
        this.baker = new BlitzSurvivalGamesKit(data, 'baker');
        this.knight = new BlitzSurvivalGamesKit(data, 'knight');
        this.guardian = new BlitzSurvivalGamesKit(data, 'guardian');
        this.scout = new BlitzSurvivalGamesKit(data, 'scout');
        this.hunter = new BlitzSurvivalGamesKit(data, 'hunter');
        this.hypeTrain = new BlitzSurvivalGamesKit(data, 'hype train');
        this.fisherman = new BlitzSurvivalGamesKit(data, 'fisherman');
        this.armorer = new BlitzSurvivalGamesKit(data, 'armorer');
        this.horsetamer = new BlitzSurvivalGamesKit(data, 'horsetamer');
        this.astronaut = new BlitzSurvivalGamesKit(data, 'astronaut');
        this.troll = new BlitzSurvivalGamesKit(data, 'troll');
        this.reaper = new BlitzSurvivalGamesKit(data, 'reaper');
        this.shark = new BlitzSurvivalGamesKit(data, 'shark');
        this.reddragon = new BlitzSurvivalGamesKit(data, 'reddragon');
        this.toxicologist = new BlitzSurvivalGamesKit(data, 'toxicologist');
        this.rogue = new BlitzSurvivalGamesKit(data, 'rogue');
        this.warlock = new BlitzSurvivalGamesKit(data, 'warlock');
        this.slimeyslime = new BlitzSurvivalGamesKit(data, 'slimeyslime');
        this.jockey = new BlitzSurvivalGamesKit(data, 'jockey');
        this.golem = new BlitzSurvivalGamesKit(data, 'golem');
        this.viking = new BlitzSurvivalGamesKit(data, 'viking');
        this.shadowKnight = new BlitzSurvivalGamesKit(data, 'shadow knight');
        this.pigman = new BlitzSurvivalGamesKit(data, 'pigman');
        this.paladin = new BlitzSurvivalGamesKit(data, 'paladin');
        this.necromancer = new BlitzSurvivalGamesKit(data, 'necromancer');
        this.florist = new BlitzSurvivalGamesKit(data, 'florist');
        this.diver = new BlitzSurvivalGamesKit(data, 'diver');
        this.arachnologist = new BlitzSurvivalGamesKit(data, 'arachnologist');
        this.blaze = new BlitzSurvivalGamesKit(data, 'blaze');
        this.wolftamer = new BlitzSurvivalGamesKit(data, 'wolftamer');
        this.tim = new BlitzSurvivalGamesKit(data, 'tim');
        this.farmer = new BlitzSurvivalGamesKit(data, 'farmer');
        this.creepertamer = new BlitzSurvivalGamesKit(data, 'creepertamer');
        this.snowman = new BlitzSurvivalGamesKit(data, 'snowman');
    }
}
export default BlitzSurvivalGames;
