// hypixel-api-reborn/dist/Structures/MiniGames/Arcade/ZombiesStats.js
class ZombiesStats {
    bestRound;
    deaths;
    doorsOpened;
    fastestRound10;
    fastestRound20;
    fastestRound30;
    playersRevived;
    timesKnockedDown;
    roundsSurvived;
    windowsRepaired;
    wins;
    zombieKills;
    constructor(data, map) {
        const mapName = map ? `_${map}` : '';
        this.bestRound = data?.[`best_round_zombies${mapName}`] || 0;
        this.deaths = data?.[`deaths_zombies${mapName}`] || 0;
        this.doorsOpened = data?.[`doors_opened_zombies${mapName}`] || 0;
        this.fastestRound10 = data?.[`fastest_time_10_zombies${mapName}_normal`] || 0;
        this.fastestRound20 = data?.[`fastest_time_20_zombies${mapName}_normal`] || 0;
        this.fastestRound30 = data?.[`fastest_time_30_zombies${mapName}_normal`] || 0;
        this.playersRevived = data?.[`players_revived_zombies${mapName}`] || 0;
        this.timesKnockedDown = data?.[`times_knocked_down_zombies${mapName}`] || 0;
        this.roundsSurvived = data?.[`total_rounds_survived_zombies${mapName}`] || 0;
        this.windowsRepaired = data?.[`windows_repaired_zombies${mapName}`] || 0;
        this.wins = data?.[`wins_zombies${mapName}`] || 0;
        this.zombieKills = data?.[`zombie_kills_zombies${mapName}`] || 0;
    }
}
export default ZombiesStats;
