// farming-weight/dist/util/jacob.js
export function formatJacobContests(contests) {
    return contests.map((contest) => {
        return {
            collected: contest.collected,
            position: contest.claimed_position,
            participants: contest.claimed_participants,
            medal: calculateJacobContestMedal(contest),
        };
    });
}
export function calculateJacobContestMedal(contest) {
    if (contest.claimed_medal) {
        return contest.claimed_medal;
    }
    const { claimed_position: position, claimed_participants: participants } = contest;
    if (position === undefined || participants === undefined) {
        return undefined;
    }
    if (position <= Math.floor(participants * 0.02))
        return 'diamond';
    if (position <= Math.floor(participants * 0.05))
        return 'platinum';
    if (position <= Math.floor(participants * 0.1))
        return 'gold';
    if (position <= Math.floor(participants * 0.3))
        return 'silver';
    if (position <= Math.floor(participants * 0.6))
        return 'bronze';
    return undefined;
}
