// hypixel-api-reborn/dist/Structures/MiniGames/BedWars/BedWarsMode.js
import BedWarsBeds from './BedWarsBeds.js';
import Divide from '../../../Utils/Divide.js';
class BedWarsMode {
    mode;
    winstreak;
    playedGames;
    kills;
    deaths;
    wins;
    losses;
    finalKills;
    finalDeaths;
    beds;
    KDR;
    WLR;
    FKDR;
    constructor(data, mode) {
        this.mode = mode;
        this.winstreak = data?.[`${mode}_winstreak`] || 0;
        this.playedGames = data?.[`${mode}_games_played_bedwars`] || 0;
        this.kills = data?.[`${mode}_kills_bedwars`] || 0;
        this.deaths = data?.[`${mode}_deaths_bedwars`] || 0;
        this.wins = data?.[`${mode}_wins_bedwars`] || 0;
        this.losses = data?.[`${mode}_losses_bedwars`] || 0;
        this.finalKills = data?.[`${mode}_final_kills_bedwars`] || 0;
        this.finalDeaths = data?.[`${mode}_final_deaths_bedwars`] || 0;
        this.beds = new BedWarsBeds(data, mode);
        this.KDR = Divide(data?.[`${mode}_kills_bedwars`], data?.[`${mode}_deaths_bedwars`]);
        this.WLR = Divide(data?.[`${mode}_wins_bedwars`], data?.[`${mode}_losses_bedwars`]);
        this.FKDR = Divide(data?.[`${mode}_final_kills_bedwars`], data?.[`${mode}_final_deaths_bedwars`]);
    }
}
export default BedWarsMode;
