// hypixel-api-reborn/dist/Structures/MiniGames/Quakecraft/Quakecraft.js
import Divide from '../../../Utils/Divide.js';
import QuakecraftMode from './QuakecraftMode.js';
class Quakecraft {
    coins;
    solo;
    teams;
    wins;
    kills;
    deaths;
    KDR;
    killstreaks;
    distanceTravelled;
    shotsFired;
    headshots;
    instantRespawn;
    killPrefixColor;
    showPrefix;
    killSound;
    barrel;
    case;
    muzzle;
    sight;
    trigger;
    constructor(data) {
        this.coins = data?.coins || data?.tokens || 0;
        this.solo = new QuakecraftMode(data);
        this.teams = new QuakecraftMode(data, 'teams');
        this.wins = this.solo?.wins + this.teams?.wins;
        this.kills = this.solo?.kills + this.teams?.kills;
        this.deaths = this.solo?.deaths + this.teams?.deaths;
        this.KDR = Divide(this.kills, this.deaths);
        this.killstreaks = this.solo?.killstreaks + this.teams?.killstreaks;
        this.distanceTravelled = this.solo?.distanceTravelled + this.teams?.distanceTravelled;
        this.shotsFired = this.solo?.shotsFired + this.teams?.shotsFired;
        this.headshots = this.solo?.headshots + this.teams?.headshots;
        this.instantRespawn = data?.instantRespawn || false;
        this.killPrefixColor = data?.selectedKillPrefix || '';
        this.showPrefix = data?.showKillPrefix || false;
        this.killSound = data?.killsound || 'None';
        this.barrel = data?.barrel || 'None';
        this.case = data?.case || 'None';
        this.muzzle = data?.muzzle || 'None';
        this.sight = data?.sight || 'None';
        this.trigger = data?.trigger || 'None';
    }
}
export default Quakecraft;
